src/web/nextui/src/app/eval/ResultsCharts.tsx

```diff
@@ -415,13 +419,238 @@ function MetricChart({ table }: ChartProps) {
   return <canvas ref={metricCanvasRef} style={{ maxHeight: '300px' }}></canvas>;
 }
 
+function LineChart({ table, evalId, config }: ChartProps) {
+  const lineCanvasRef = useRef(null);
+  const lineChartInstance = useRef<Chart | null>(null);
+  const [datasetId, setDatasetId] = useState('');
+  const [dataset, setDataset] = useState<TestCasesWithMetadata>();
+
+  interface PointMetadata {
+    evalId: string;
+    highlight: boolean;
+    label: string;
+    date: string;
+    score: string;
+  }
+
+  interface Point {
+    x: number;
+    y: number;
+    metadata: PointMetadata;
+  }
+
+  function ordinalSuffixOf(i: number): string {
+    const j = i % 10,
+      k = i % 100;
+    if (j === 1 && k !== 11) {
+      return i + 'st';
+    }
+    if (j === 2 && k !== 12) {
+      return i + 'nd';
+    }
+    if (j === 3 && k !== 13) {
+      return i + 'rd';
+    }
+    return i + 'th';
+  }
+
+  useEffect(() => {
+    if (config) {
+      const newDatasetId = createHash('sha256').update(JSON.stringify(config.tests)).digest('hex');
+      setDatasetId(newDatasetId);
```

Comment:
As `setDatasetId` is only called once (this `useEffect`-scoped initializer), you can reduce the `useState` and `useEffect` into a single `[useMemo](https://react.dev/reference/react/useMemo)`.

Reply:
Thank you for your advice. Fixed!